ddl/foreign_key.go

```diff
@@ -112,3 +114,143 @@ func onDropForeignKey(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ err
 		return ver, dbterror.ErrInvalidDDLState.GenWithStackByArgs("foreign key", fkInfo.State)
 	}
 }
+
+func allocateFKIndexID(tblInfo *model.TableInfo) int64 {
+	tblInfo.MaxForeignKeyID++
+	return tblInfo.MaxForeignKeyID
+}
+
+func checkTableForeignKeysValid(sctx sessionctx.Context, is infoschema.InfoSchema, schema string, tbInfo *model.TableInfo) error {
+	if !sctx.GetSessionVars().EnableForeignKey {
+		return nil
+	}
+	fkCheck := sctx.GetSessionVars().ForeignKeyChecks
+	for _, fk := range tbInfo.ForeignKeys {
+		if fk.Version < 1 {
+			continue
+		}
+		err := checkTableForeignKeyValid(is, schema, tbInfo, fk, fkCheck)
+		if err != nil {
+			return err
+		}
+	}
+
+	referredFKInfos := is.GetTableReferredForeignKeys(schema, tbInfo.Name.L)
+	for _, referredFK := range referredFKInfos {
+		childTable, err := is.TableByName(referredFK.ChildSchema, referredFK.ChildTable)
+		if err != nil {
+			return err
+		}
+		fk := model.FindFKInfoByName(childTable.Meta().ForeignKeys, referredFK.ChildFKName.L)
+		if fk == nil {
+			continue
+		}
+		err = checkTableForeignKey(tbInfo, childTable.Meta(), fk)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func checkTableForeignKeyValid(is infoschema.InfoSchema, schema string, tbInfo *model.TableInfo, fk *model.FKInfo, fkCheck bool) error {
+	var referTblInfo *model.TableInfo
+	if fk.RefSchema.L == schema && fk.RefTable.L == tbInfo.Name.L {
+		same := true
+		for i, col := range fk.Cols {
+			if col.L != fk.RefCols[i].L {
+				same = false
+				break
+			}
+		}
+		if same {
+			// self-reference with same columns is not support.
+			return infoschema.ErrCannotAddForeign
+		}
+		referTblInfo = tbInfo
+	} else {
+		referTable, err := is.TableByName(fk.RefSchema, fk.RefTable)
+		if err != nil {
+			if (infoschema.ErrTableNotExists.Equal(err) || infoschema.ErrDatabaseNotExists.Equal(err)) && !fkCheck {
+				return nil
+			}
+			return infoschema.ErrFkCannotOpenParent.GenWithStackByArgs(fk.RefTable.O)
+		}
+		referTblInfo = referTable.Meta()
+	}
+	return checkTableForeignKey(referTblInfo, tbInfo, fk)
+}
+
+func checkTableForeignKeyValidInOwner(d *ddlCtx, job *model.Job, tbInfo *model.TableInfo, fkCheck bool) error {
+	is := d.infoCache.GetLatest()
```

Comment:
It's better to use `meta.Meta` to get a table's definition to make sure all read/write in a txn are based on a same version. However, it seems a little hard to `GetTableReferredForeignKeys` using `meta.Meta`. Can you add some comments for how the meta keeps consistency here?

Reply:
Nice catch. Since the `d.infoCache.GetLatest()` maybe not the latest,  so use `meta.Meta` to get table info is better.

Reply:
Nice catch, I need to make sure the `d.infoCache.GetLatest` is really the latest version schema. So I add a check here now